### PR TITLE
New Zeus script for all langages.

### PR DESCRIPTION
**Pull request by mtancoigne**
Updated the zeus script. Don't attach any importance at the branch name, nothing's related to PHP beside the fact I only tested with my box.

So the big changes : 
- Zeus script is called `zeus -l <language> -d <destination> -o <option>`
- Every repo should have a `/vagrant/vagrant/_provision_vars.sh` with at least `export V_WORKING_DIR='some/path'`, the path being the absolute path where the git repo must be initialized.
- You can use the `-o` option to create variations on your provisionning (look at [how I did it](https://github.com/alayek/zeus/blob/PHP-zeus-script-test/php/vagrant/provision.sh#L115)
- In the `/vagrant/vagrant/` dir of the created box, a file named `zeus_credentials.sh` is copied. It contains the script to authenticate to github/heroku and init a git repo in the `V_WORKING_DIR`

I think that's all the changes, but for now you don't have to bother with implementing a .bashrc to achieve this.

@alayek Maybe the best way to merge this without breaking anything would be to merge the PHP branch first, as this one is a fork.

### CODE REVIEW

**Comment by mtancoigne** [comment]
Tested with #58, works well for 
- `zeus -l php -d <dir>` (base PHP setup)
- `zeus -l php -d <dir> -o cake3` (PHP+CakePHP)

**Comment by alayek** [comment]
At present, this works well for PHP. But other directories need to have that structure for their provisioning files. And remember it's not just the provisioning script - if the script copies files into vagrant box, then they need to be appropriately dealt with as well.

---

I don't like the added complexity. Earlier it was just `vagrant up`. In worst case, it would have been `cd <language> && vagrant up`.

But now, one needs to provide a few extra stuff. I think the rules are rigid.

So we need to make the user experience better.

Here are my proposals - 
- If a user simply provides `./zeus` at command line, it could ask like this:

``` bash
$ ./zeus
Which language would you like to work on? (default: node, type Ctrl + C to exit) >>
```

The user might enter PHP or PhP or Php - the matching should be case-insensitive. Also, resolve ambiguity - if someone types node or Node JS - it would work the same; which means it has to be a content-search, not a start-with type of search.

If the user enters something the script couldn't match, prompt the user again saying you didn't understand.
- If the user doesn't provide a directory, make a directory `fcc_<language>` in home directory of the user. Also pick auto-correcting, like we have for port forwarding. If there is already a non-empty directory, add a number at the end end of directory name (I am sure there are such autocorrect functionality) and inform the user.

Don't error out!

The prompt could look like this:

``` bash
$ ./zeus
Which language would you like to work on? (default: node, hit enter to accept, type Ctrl + C to exit) >> Php
In which directory, should we create the workspace? (default: fcc_php, hit enter to accept, type Ctrl + C to exit) >>
```

Also, if previous format is used by someone - that should work too.

When the zeus provisiooning ends, the control is send back to the script; so take the user to his workspace with a `cd` command and tell him to do `vagrant ssh` from there.

This shows maturity on part of the script creator. I know it's not easy to build such a thing; but we should aspire to build this!!

**Comment by mtancoigne** [comment]
@alayek the -d argument is the directory where to put the vagrant box. For a typical use:
- `git clone https://github.com/alayek/zeus zeus` will clone in zeus folder.
- then `cd zeus`.
- Let's say I want to create a python box in my `home/boxes/` folder, i would call `zeus -l python -d ~/boxes/fcc-python`. When the box is up, I only have to vagrant ssh in it and use it.

The _extra stuff_, as you say, is only for the box _makers_. I don't put all my code in the root of the shared `/vagrant` folder, that's a nonsense, as I only want to commit actual code. So for the PHP boxes, there's a `/vagrant/www` folder. And when the zeus script sets up a git repo, it will not make it a the root of the vagrant box, but in the www folder. So this folder must be specified. But the user don't have anything to to, it's all to the box maker to write in a file where his working dir is. And all this depends of the language we use.

I understand you can run a python script from almost everywhere, but in the case of a server, things have to be tightly set up...

The zeus script i wrote only copy a skeleton box where the user wants it to be and allow box devs to add a bit more of control. Then, the user don't need zeus anymore, his box is a standalone box he can raise with vagrant up, as the config files has been written during the box copy process... 

And there is an other script, copied IN the box, to configure/create git/Heroku repos directly in the space the box dev assigned to the user.

If that sounds too complicated, I understand it, but keep this PR for later, i'm sure I can make you change your mind :)

And I think a certain misunderstanding has been made by me in my hurry to unify the different languages boxes with a single script, as I think @byteknacker's zeus is in the PATH and allows to  create git repos on the fly in the whatever current folder, as mine only creates a git repo where the box dev thought it was the place to make it.

**Comment by alayek** [comment]
Exactly. Taking the `git` example, `git clone` doesn't ask you where to clone.

It picks a value, and goes with it. It the directory is non-empty, it tells user that it cannot clone. If you provide one more parameter, it takes that a directory to clone into.

---

The best interface is no interface. If you make user provide input to anything, be sure you handle errors.

A `-d` option is good; but also make it so that a user doesn't have to specify and it still picks something default.

**Comment by byteknacker** [comment]
Ok I think I will give a lot of thought into the commandline structure and give me proposals tomorrow on gitter and create some documentation and flow charts. But in principle my suggestions would be:
1. Unified syntax for only one single file: zeus which rules them all
2. No different bash scripts for different purposes but rather if any, then they act as libraries of separated functions all being called by a main script: zeus. So you have zeus-php, zeus-python etc but these only contain classes and functions. Each contributor for a particular stack can just focus on coding this part of zeus. The main zeus file shall be unchanged as long as possible. 
3. Adopt a habit of making a code more general and with the goal not to necessitate a change just because of some new use cases. It should be powerful enough to capture them.
4. Simple is better than complicated. That means, we may consider juxtaposition and default values and perhaps also interactive input prompts for users to have a convenience of just using one command line argument to do stuff they want to do. 
5. All zeus commands should be available on the local machine and in every single stack VM so generated. This is because people don't have to learn ten different commandline syntax for ten different stack. Imagine we specialised for PHP with a different, even just a little different command line syntax, after a few weeks, we have twenty or so stacks, each of which has slightly different commandline syntax, this would be a nightmare for the user to learn and they will not use all of the capabilities of zeus. Hence, we should preserve zeus for all stack and make it polymorph, behave differently depending on the situation and stack but extremly uniform in the syntax call.

**Comment by byteknacker** [comment]
@mtancoigne I definitely welcome your changes proposed and I will take in all the goodies from your new commits. But I do think, being honest, that we need to make some upgrades in terms of coherence of zeus across the entire project, meaning agree all of us on how exactly we wish to structure the command line syntax and in the future the public API.

File: php/.gitignore [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+**/*.pot
+**/local_settings.py
+**/db.sqlite3
+**/media/
+
+### Vagrant Directory###
+.vagrant/
+**/.vagrant/
+
+### Custom Directories###
```
Comment (ghost) [inline php/.gitignore:67]: ~~### Custom Directories###~~ ### Custom Directories ###

**Comment by byteknacker** [comment]
@atjonathan are you using Grammarly?

**Comment by ghost** [comment]
@byteknacker, no just manual scanning!

**Comment by byteknacker** [comment]
@mtancoigne would you mind to change the destination of your PR from DevBlend:master to DevBlend:dev?

**Comment by ghost** [comment]
Closing. Made a new PR (#75) that targets the dev branch.

**Comment by byteknacker** [comment]
Many thanks!

**Pull request closed without merging**